server/schedulers/hot_region.go

```diff
@@ -606,184 +660,258 @@ func (bs *balanceSolver) getDstCandidateIDs() map[uint64]struct{} {
 			filter.NewHealthFilter(bs.sche.GetName()),
 		}
 
-		candidates = bs.cluster.GetFollowerStores(bs.region)
+		candidates = bs.cluster.GetFollowerStores(bs.cur.region)
 
 	default:
 		return nil
 	}
 
-	ret := make(map[uint64]struct{}, len(candidates))
+	ret := make(map[uint64]*storeLoadDetail, len(candidates))
 	for _, store := range candidates {
 		if !filter.Target(bs.cluster, store, filters) {
-			ret[store.GetID()] = struct{}{}
+			ret[store.GetID()] = bs.stLoadDetail[store.GetID()]
 		}
 	}
 	return ret
 }
 
-func (bs *balanceSolver) selectDstStoreID(candidateIDs map[uint64]struct{}) uint64 {
-	candidateLoadDetail := make(map[uint64]*storeLoadDetail, len(candidateIDs))
-	for id := range candidateIDs {
-		candidateLoadDetail[id] = bs.stLoadDetail[id]
-	}
-	switch bs.opTy {
-	case movePeer:
-		return selectDstStoreByByteRate(candidateLoadDetail, bs.srcPeerStat.GetByteRate(), bs.stLoadDetail[bs.srcStoreID])
-	case transferLeader:
-		if bs.rwTy == write {
-			return selectDstStoreByCount(candidateLoadDetail, bs.srcPeerStat.GetByteRate(), bs.stLoadDetail[bs.srcStoreID])
+// calcProgressiveRank checks `bs.cur` is a progressive solution.
+func (bs *balanceSolver) calcProgressiveRank() int64 {
+	srcLd := bs.stLoadDetail[bs.cur.srcStoreID].LoadPred.min()
+	dstLd := bs.stLoadDetail[bs.cur.dstStoreID].LoadPred.max()
+	peer := bs.cur.srcPeerStat
+	rank := int64(0)
+	if bs.rwTy == write && bs.opTy == transferLeader {
+		if srcLd.Count > dstLd.Count &&
+			srcLd.KeyRate >= dstLd.KeyRate+peer.GetKeyRate() {
+			rank = -1
+		}
+	} else {
+		keyDecRatio := (dstLd.KeyRate + peer.GetKeyRate()) / (srcLd.KeyRate + 1)
+		keyHot := peer.GetKeyRate() >= 10
```

Comment:
how about using a constant?

Reply:
It won't be used in other place, so I don't think it is necessary.

Reply:
The constant is easy to maintain and the meaning is much more clear.

Reply:
Add two constants: `minHotByteRate` and `minHotKeyRate`.